# Price Cut Combined Random Forest

### Data Cleanup

The following property types were eliminated from the original CSV:
1. Lot/Land

In [1]:
# Dependencies
import pandas as pd
from pathlib import Path

In [2]:
# Name of the CSV file
file_path = Path("Resources/For_Sale_Data_Combined.csv")
df = pd.read_csv(file_path)
df

Property URL  MLS ID  \
0     https://www.zillow.com/homedetails/93-N-Cooper...     NaN   
1     https://www.zillow.com/homedetails/295-N-Rural...     NaN   
2     https://www.zillow.com/homedetails/1351-N-Plea...     NaN   
3     https://www.zillow.com/homedetails/1075-E-Chan...     NaN   
4     https://www.zillow.com/homedetails/295-N-Rural...     NaN   
...                                                 ...     ...   
1680  https://www.zillow.com/homedetails/8812-SE-Rur...     NaN   
1681  https://www.zillow.com/homedetails/5421-NE-Kil...     NaN   
1682  https://www.zillow.com/homedetails/1333-N-Albe...     NaN   
1683  https://www.zillow.com/homedetails/1524-NE-Mar...     NaN   
1684  https://www.zillow.com/homedetails/15061-NW-Eu...     NaN   

     Listing description Property type  Number of days on Zillow  \
0     Apartment for sale     Apartment                       NaN   
1     Apartment for sale     Apartment                       5.0   
2     Apartment for sale     Apartment                       4.0   
3     Apartment for sale     Apartment                      15.0   
4     Apartment for sale     Apartment                       NaN   
...                  ...           ...                       ...   
1680              Active       Unknown                     234.0   
1681              Active       Unknown                       8.0   
1682              Active       Unknown                       8.0   
1683              Active       Unknown                       9.0   
1684              Active       Unknown                       NaN   

      Property price (USD) Price was cut  Price cut amount (USD)  \
0                 289000.0            No                     NaN   
1                 389999.0            No                     NaN   
2                 398000.0            No                     NaN   
3                 225000.0            No                     NaN   
4                 299900.0           Yes                 35100.0   
...                    ...           ...                     ...   
1680              429000.0           Yes                  6000.0   
1681             1000000.0            No                     NaN   
1682              585000.0            No                     NaN   
1683              599000.0            No                     NaN   
1684             1150000.0           Yes                450000.0   

     Price cut date (MM/DD/YYYY)  Living area  ... Bedrooms  Bathrooms  \
0                            NaN       1075.0  ...      2.0        2.0   
1                            NaN       1069.0  ...      2.0        2.0   
2                            NaN       1480.0  ...      3.0        2.0   
3                            NaN        705.0  ...      1.0        1.0   
4                     02/13/2024        979.0  ...      1.0        1.0   
...                          ...          ...  ...      ...        ...   
1680                  07/25/2023       1270.0  ...      3.0        2.0   
1681                         NaN       3328.0  ...      8.0        4.0   
1682                         NaN       1850.0  ...      2.0        2.0   
1683                         NaN       2273.0  ...      2.0        2.0   
1684                  01/25/2024       3276.0  ...      5.0        4.0   

                          Broker name Broker agent  \
0                    RE/MAX Solutions          NaN   
1               RE/MAX Alliance Group          NaN   
2                        DeLex Realty          NaN   
3                      DPR Realty LLC          NaN   
4            Arizona Elite Properties          NaN   
...                               ...          ...   
1680              Sundance Realty LLC          NaN   
1681               Branch Real Estate          NaN   
1682               Change Realty, LLC          NaN   
1683     John L. Scott Portland South          NaN   
1684  Keller Williams Sunset Corridor          NaN   

                     Street address      City    Zip State Country

In [3]:
# Clean the dataframe by replacing NaN values with 0
df.fillna(0, inplace=True)

In [4]:
df

Property URL  MLS ID  \
0     https://www.zillow.com/homedetails/93-N-Cooper...     0.0   
1     https://www.zillow.com/homedetails/295-N-Rural...     0.0   
2     https://www.zillow.com/homedetails/1351-N-Plea...     0.0   
3     https://www.zillow.com/homedetails/1075-E-Chan...     0.0   
4     https://www.zillow.com/homedetails/295-N-Rural...     0.0   
...                                                 ...     ...   
1680  https://www.zillow.com/homedetails/8812-SE-Rur...     0.0   
1681  https://www.zillow.com/homedetails/5421-NE-Kil...     0.0   
1682  https://www.zillow.com/homedetails/1333-N-Albe...     0.0   
1683  https://www.zillow.com/homedetails/1524-NE-Mar...     0.0   
1684  https://www.zillow.com/homedetails/15061-NW-Eu...     0.0   

     Listing description Property type  Number of days on Zillow  \
0     Apartment for sale     Apartment                       0.0   
1     Apartment for sale     Apartment                       5.0   
2     Apartment for sale     Apartment                       4.0   
3     Apartment for sale     Apartment                      15.0   
4     Apartment for sale     Apartment                       0.0   
...                  ...           ...                       ...   
1680              Active       Unknown                     234.0   
1681              Active       Unknown                       8.0   
1682              Active       Unknown                       8.0   
1683              Active       Unknown                       9.0   
1684              Active       Unknown                       0.0   

      Property price (USD) Price was cut  Price cut amount (USD)  \
0                 289000.0            No                     0.0   
1                 389999.0            No                     0.0   
2                 398000.0            No                     0.0   
3                 225000.0            No                     0.0   
4                 299900.0           Yes                 35100.0   
...                    ...           ...                     ...   
1680              429000.0           Yes                  6000.0   
1681             1000000.0            No                     0.0   
1682              585000.0            No                     0.0   
1683              599000.0            No                     0.0   
1684             1150000.0           Yes                450000.0   

     Price cut date (MM/DD/YYYY)  Living area  ... Bedrooms  Bathrooms  \
0                              0       1075.0  ...      2.0        2.0   
1                              0       1069.0  ...      2.0        2.0   
2                              0       1480.0  ...      3.0        2.0   
3                              0        705.0  ...      1.0        1.0   
4                     02/13/2024        979.0  ...      1.0        1.0   
...                          ...          ...  ...      ...        ...   
1680                  07/25/2023       1270.0  ...      3.0        2.0   
1681                           0       3328.0  ...      8.0        4.0   
1682                           0       1850.0  ...      2.0        2.0   
1683                           0       2273.0  ...      2.0        2.0   
1684                  01/25/2024       3276.0  ...      5.0        4.0   

                          Broker name Broker agent  \
0                    RE/MAX Solutions            0   
1               RE/MAX Alliance Group            0   
2                        DeLex Realty            0   
3                      DPR Realty LLC            0   
4            Arizona Elite Properties            0   
...                               ...          ...   
1680              Sundance Realty LLC            0   
1681               Branch Real Estate            0   
1682               Change Realty, LLC            0   
1683     John L. Scott Portland South            0   
1684  Keller Williams Sunset Corridor            0   

                     Street address      City    Zip State Country

In [5]:
# Import scikit-learn libraries including Random Forest Classifier
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier  # Changed from DecisionTreeClassifier to RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import plot_tree
import matplotlib.pyplot as plt

In [6]:
# Convert 'Price was cut' to a binary indicator (0 and 1)
df['Price was cut'] = df['Price was cut'].apply(lambda x: 1 if x == 'Yes' else 0)
df.tail()

Property URL  MLS ID  \
1680  https://www.zillow.com/homedetails/8812-SE-Rur...     0.0   
1681  https://www.zillow.com/homedetails/5421-NE-Kil...     0.0   
1682  https://www.zillow.com/homedetails/1333-N-Albe...     0.0   
1683  https://www.zillow.com/homedetails/1524-NE-Mar...     0.0   
1684  https://www.zillow.com/homedetails/15061-NW-Eu...     0.0   

     Listing description Property type  Number of days on Zillow  \
1680              Active       Unknown                     234.0   
1681              Active       Unknown                       8.0   
1682              Active       Unknown                       8.0   
1683              Active       Unknown                       9.0   
1684              Active       Unknown                       0.0   

      Property price (USD)  Price was cut  Price cut amount (USD)  \
1680              429000.0              1                  6000.0   
1681             1000000.0              0                     0.0   
1682              585000.0              0                     0.0   
1683              599000.0              0                     0.0   
1684             1150000.0              1                450000.0   

     Price cut date (MM/DD/YYYY)  Living area  ... Bedrooms  Bathrooms  \
1680                  07/25/2023       1270.0  ...      3.0        2.0   
1681                           0       3328.0  ...      8.0        4.0   
1682                           0       1850.0  ...      2.0        2.0   
1683                           0       2273.0  ...      2.0        2.0   
1684                  01/25/2024       3276.0  ...      5.0        4.0   

                          Broker name Broker agent            Street address  \
1680              Sundance Realty LLC            0          8812 SE Rural St   
1681               Branch Real Estate            0  5421 NE Killingsworth St   
1682               Change Realty, LLC            0         1333 N Alberta St   
1683     John L. Scott Portland South            0         1524 NE Marine Dr   
1684  Keller Williams Sunset Corridor            0        15061 NW Eugene Ln   

          City    Zip State Country FederalInt  
1680  Portland  97266    OR     USA       5.33  
1681  Portland  97218    OR     USA       5.33  
1682  Portland  97217    OR     USA       5.33  
1683  Portland  97211    OR     USA       5.33  
1684  Portland  97229    OR     USA       5.33  

[5 rows x 24 columns]

In [7]:
# Feature selection
# Selecting relevant features and the target variable
X = df[['Property price (USD)', 'Property type', 'Living area', 'Lot/land area', 'Bedrooms', 'Bathrooms', 'City', 'FederalInt']]
y = df['Price was cut']


In [8]:
# Handle the categorical variables
categorical_features = ['Property type', 'City']
label_encoders = {}

for col in categorical_features:
    le = LabelEncoder()
    X.loc[:,col] = le.fit_transform(X[col])
    label_encoders[col] = le

In [9]:
df

Property URL  MLS ID  \
0     https://www.zillow.com/homedetails/93-N-Cooper...     0.0   
1     https://www.zillow.com/homedetails/295-N-Rural...     0.0   
2     https://www.zillow.com/homedetails/1351-N-Plea...     0.0   
3     https://www.zillow.com/homedetails/1075-E-Chan...     0.0   
4     https://www.zillow.com/homedetails/295-N-Rural...     0.0   
...                                                 ...     ...   
1680  https://www.zillow.com/homedetails/8812-SE-Rur...     0.0   
1681  https://www.zillow.com/homedetails/5421-NE-Kil...     0.0   
1682  https://www.zillow.com/homedetails/1333-N-Albe...     0.0   
1683  https://www.zillow.com/homedetails/1524-NE-Mar...     0.0   
1684  https://www.zillow.com/homedetails/15061-NW-Eu...     0.0   

     Listing description Property type  Number of days on Zillow  \
0     Apartment for sale     Apartment                       0.0   
1     Apartment for sale     Apartment                       5.0   
2     Apartment for sale     Apartment                       4.0   
3     Apartment for sale     Apartment                      15.0   
4     Apartment for sale     Apartment                       0.0   
...                  ...           ...                       ...   
1680              Active       Unknown                     234.0   
1681              Active       Unknown                       8.0   
1682              Active       Unknown                       8.0   
1683              Active       Unknown                       9.0   
1684              Active       Unknown                       0.0   

      Property price (USD)  Price was cut  Price cut amount (USD)  \
0                 289000.0              0                     0.0   
1                 389999.0              0                     0.0   
2                 398000.0              0                     0.0   
3                 225000.0              0                     0.0   
4                 299900.0              1                 35100.0   
...                    ...            ...                     ...   
1680              429000.0              1                  6000.0   
1681             1000000.0              0                     0.0   
1682              585000.0              0                     0.0   
1683              599000.0              0                     0.0   
1684             1150000.0              1                450000.0   

     Price cut date (MM/DD/YYYY)  Living area  ... Bedrooms  Bathrooms  \
0                              0       1075.0  ...      2.0        2.0   
1                              0       1069.0  ...      2.0        2.0   
2                              0       1480.0  ...      3.0        2.0   
3                              0        705.0  ...      1.0        1.0   
4                     02/13/2024        979.0  ...      1.0        1.0   
...                          ...          ...  ...      ...        ...   
1680                  07/25/2023       1270.0  ...      3.0        2.0   
1681                           0       3328.0  ...      8.0        4.0   
1682                           0       1850.0  ...      2.0        2.0   
1683                           0       2273.0  ...      2.0        2.0   
1684                  01/25/2024       3276.0  ...      5.0        4.0   

                          Broker name Broker agent  \
0                    RE/MAX Solutions            0   
1               RE/MAX Alliance Group            0   
2                        DeLex Realty            0   
3                      DPR Realty LLC            0   
4            Arizona Elite Properties            0   
...                               ...          ...   
1680              Sundance Realty LLC            0   
1681               Branch Real Estate            0   
1682               Change Realty, LLC            0   
1683     John L. Scott Portland South            0   
1684  Keller Williams Sunset Corridor            0   

                     Street address      City    Zip S

In [10]:
# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [11]:
# Initializing and training the Decision Tree Classifier
clf = RandomForestClassifier(random_state=42)
clf.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [12]:
# Make predictions
y_pred = clf.predict(X_test)


In [13]:
# Evaluating the model
print("Accuracy Score:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))

Accuracy Score: 0.6261127596439169
Confusion Matrix:
 [[174  48]
 [ 78  37]]
Classification Report:
               precision    recall  f1-score   support

           0       0.69      0.78      0.73       222
           1       0.44      0.32      0.37       115

    accuracy                           0.63       337
   macro avg       0.56      0.55      0.55       337
weighted avg       0.60      0.63      0.61       337



# Initial Result of the Random Forest Model

### Accuracy Score
The accuracy score of 0.626 (or 62.6%) indicates that the model correctly predicts the outcome (whether the price will be cut or not) in approximately 62.6% of the cases. This metric provides a general idea of the model's overall performance but doesn't detail its effectiveness in distinguishing between the two classes.



In [14]:
# Note: Random Forest trees might be too complex for practical visualization with plot_tree,
# especially for large numbers of trees. Consider summarizing tree characteristics or feature importances instead.
feature_importances = pd.Series(clf.feature_importances_, index=X.columns).sort_values(ascending=False)
print(feature_importances)

Property price (USD)    0.291376
Living area             0.275927
Lot/land area           0.193990
Bedrooms                0.086842
Bathrooms               0.067339
City                    0.047285
Property type           0.037242
FederalInt              0.000000
dtype: float64
